### DISCUSSION TITLE
Why the Cookies API isn't available on the client?

### DISCUSSION THREAD

**Discussion by julianpera** [vercel/next.js#51116] [App Router]
Some fetch calls on my application rely on cookies for authentication, but because `cookies()` is not available on the client, I have to write two implementation of the same function: one for the server and one for the client.

I tried an idea for a workaround, checking if `typeof(window) === 'undefined'`, so the function could behave differently based on the environment it's running, but NextJS complains when I call it on the client because of `cookies()`, even through `cookies()` is not being invoked at all.

```ts
async function getProducts() {
  const accessToken = typeof(window) === 'undefined' ? cookies().get('access_token') : parseDocumentCookie('access_token');
  const response = await fetch(PRODUCTS_URL, {
    headers: {
      Authorization: `Bearer ${accessToken}`,
    },
  });
  const products = await response.json();
  return products;
}
```

If cookies are available on the client, why the Cookies API isn't?
Any workaround to avoid writing two implementations of the same fetch?

**Comment by icyJoseph** [comment]
Cookies are special because, they might be HTTPOnly, in that case JavaScript won't be able to access them. A client side cookies function would create the wrong expectation around the abstraction, because not all cookies are bound to exist. 

Moreover, the JS available cookies exist in the document so you can always just read it from there, as you have done, it is a solved problem. 

Whereas the within a Server Component, we don't have access to the underlying request, and so, cookies and header functions are made available to use instead.

You could always parametrise, `getProducts` to take the `accessToken` as an input, and then calculate the token in the given environment instead, and avoid environment checks within the function altogether. 

Also, the way you have it, will crash out on build time, whenever imported into a Client Component (`use client`) tree:

```js
export async function getProducts() {
  let accessToken;
  if (typeof window === "undefined") {
    const { cookies } = await import("next/headers");
    accessToken = cookies().get("access_token");
  } else {
    accessToken = window.document.cookie; // or the parseDocumentCookie function
  }

  const response = await fetch("https://pokeapi.co/api/v2/pokemon/ditto", {
    headers: {
      Authorization: `Bearer ${accessToken}`,
    },
  });

  const products = await response.json();
  return products;
}
```

This has some added benefits from dead code elimination.

### Server Side after code elimination

```js
;// CONCATENATED MODULE: ./src/app/lib.ts
async function getProducts() {
    let accessToken;
    if (true) {
        const { cookies  } = await __webpack_require__.e(/* import() */ 859).then(__webpack_require__.t.bind(__webpack_require__, 859, 23));
        accessToken = cookies().get("access_token");
    } else {}
    const response = await fetch("https://pokeapi.co/api/v2/pokemon/ditto", {
        headers: {
            Authorization: `Bearer ${accessToken}`
        }
    });
    const products = await response.json();
    return products;
}
```

The `else` branch with the client side code, is gone!

### Client Side after code elimination

The function gets inlined in this case! And it looks like:

```js
 async function o() {
        let e;
        e = window.document.cookie;
        let t = await fetch("https://pokeapi.co/api/v2/pokemon/ditto", {
            headers: { Authorization: "Bearer ".concat(e) },
          }),
          n = await t.json();
        return n;
      }
```

The `next/headers` branch is gone!

I guess you could also use an API route, which your client sends requests to, and in turn that API route gets the cookie and forwards the request. In this case, you'd make the cookie HTTPOnly, and hide it altogether from the JS environment.

To sum up, is not that you are in the wrong here either. The problem is that the `cookies` function server side, is there to solve a problem that's not present in the client side. The problem being that the request that brings in the `cookies` is out of reach.
